fcm_django/models.py

```diff
@@ -41,222 +45,301 @@ def get_queryset(self):
+            return e
+
+    def handle_topic_subscription(
+        self,
+        should_subscribe: bool,
+        topic: str,
+        **more_subscribe_kwargs,
+    ) -> FirebaseResponseDict:
         """
-        Send single data message.
+        Subscribes or Unsubscribes based on instance's registration_id
+
+        :param should_subscribe: whether to have these users subscribe (True) or
+        unsubscribe to a topic (False).
+        :param topic: Name of the topic to subscribe to. May contain the ``/topics/``
+        prefix.
+        :param more_subscribe_kwargs: Parameters for
+        ``firebase.messaging.subscribe_to_topic()``
+        - app: firebase_admin.App. Specify a specific app to use
+        If there are any new parameters, you can still specify them here.
+
+        :raises FirebaseError
+        :returns FirebaseResponseDict
         """
-        from .fcm import fcm_send_single_device_data_message
-        result = fcm_send_single_device_data_message(
-            registration_id=str(self.registration_id),
-            condition=condition,
-            collapse_key=collapse_key,
-            delay_while_idle=delay_while_idle,
-            time_to_live=time_to_live,
-            restricted_package_name=restricted_package_name,
-            low_priority=low_priority,
-            dry_run=dry_run,
-            data_message=data_message,
-            content_available=content_available,
-            api_key=api_key,
-            timeout=timeout,
-            json_encoder=json_encoder,
+        _r_ids = [self.registration_id]
+        response = (
+            messaging.subscribe_to_topic
+            if should_subscribe
+            else messaging.unsubscribe_from_topic
+        )(_r_ids, topic, **more_subscribe_kwargs)
+        return FirebaseResponseDict(
+            response=response,
+            registration_ids_sent=_r_ids,
+            deactivated_registration_ids=type(
+                self
+            ).objects.deactivate_devices_with_error_results(_r_ids, response.errors),
         )
 
-        self._deactivate_device_on_error_result(result)
-        return result
-
-    def _deactivate_device_on_error_result(self, result):
-        device = FCMDevice.objects.filter(registration_id=self.registration_id)
-        if 'error' in result['results'][0]:
-            error_list = ['MissingRegistration', 'MismatchSenderId', 'InvalidRegistration', 'NotRegistered']
-            if result['results'][0]['error'] in error_list:
-                device.update(active=False)
-                self._delete_inactive_device_if_requested(device)
-
-    @staticmethod
-    def _delete_inactive_device_if_requested(device):
-        if SETTINGS["DELETE_INACTIVE_DEVICES"]:
-            device.delete()
+    @classmethod
+    def deactivate_devices_with_error_result(
+        cls, registration_id, firebase_exc, name=None
+    ) -> List[str]:
+        return cls.objects.deactivate_devices_with_error_results(
```

Comment:
this line is throwing an error,
![image](https://user-images.githubusercontent.com/7256491/124380583-b9db0380-dcbd-11eb-99ea-6cc0733f7ee9.png)